rgd/geodata/permissions.py

```diff
@@ -0,0 +1,120 @@
+from django.contrib.auth.backends import BaseBackend
+from django.core.exceptions import PermissionDenied
+
+from rgd.geodata import models
+
+
+def get_collection_membership_path(model):
+    """Get the path to the 'CollectionMembership' model.
+
+    Relationships are represented as 'dunder's ('__').
+    """
+    # Collection
+    if issubclass(model, models.CollectionMembership):
+        return ''
+    if issubclass(model, models.Collection):
+        return 'collection_memberships'
+    # Common
+    if issubclass(model, models.ChecksumFile):
+        return 'collection__collection_memberships'
+    # Imagery
+    if issubclass(model, models.ImageEntry):
+        return 'image_file__collection__collection_memberships'
+    if issubclass(model, models.Thumbnail):
+        return 'image_entry__image_file__collection__collection_memberships'
+    if issubclass(model, models.ImageSet):
+        return 'images__image_file__collection__collection_memberships'
+    if issubclass(model, models.RasterEntry):
+        return 'image_set__images__image_file__collection__collection_memberships'
+    if issubclass(model, models.RasterMetaEntry):
+        return 'parent_raster__image_set__images__image_file__collection__collection_memberships'
+    if issubclass(model, models.BandMetaEntry):
+        return 'parent_image__image_file__collection__collection_memberships'
+    if issubclass(model, models.ConvertedImageFile):
+        return 'source_image__image_file__collection__collection_memberships'
+    if issubclass(model, models.SubsampledImage):
+        return 'source_image__image_file__collection__collection_memberships'
+    if issubclass(model, models.KWCOCOArchive):
+        return 'spec_file__collection__collection_memberships'
+    # Annotation
+    if issubclass(model, models.Annotation):
+        return 'image__image_file__collection__collection_memberships'
+    if issubclass(model, models.Segmentation):
+        return 'annotation__image__image_file__collection__collection_memberships'
+    # Geometry
+    if issubclass(model, models.GeometryEntry):
+        return 'geometry_archive__collection__collection_memberships'
+    # FMV
+    if issubclass(model, models.FMVEntry):
+        return 'fmv_file__collection__collection_memberships'
+    return None
+
+
+def filter_perm(user, queryset, role):
+    """Filter a queryset."""
+    # Called outside of view
+    if user is None:
+        return queryset
+    # Admins can see all
+    if user.is_active and (user.is_staff or user.is_superuser):
+        return queryset
+    # No relationship to collection
+    path = get_collection_membership_path(queryset.model)
+    if path is None:
+        return queryset
```

Comment:
> Maybe we should implement `get_collection_path` on each model class starting with a `NotImplementedError` in `ModifiableEntry`? Or is this intended for models that might not belong to a collection?

I think this is definitely the smartest way to go to ensure permissions are set. I was hoping to centralize the logic for permissions in one file as best as I could to make it clear how to manipulate them while it is a WIP. In either case, I think it's a good idea to default to `NotImplementedError` rather than `None`.

Reply:
I wanted to follow up on this. Now `get_collection_membership_path` must explicitly return `None` if the model should not be checked for permissions. Otherwise, this function defaults to `NotImplementedError`.

I think this accomplishes the behavior we're looking for here, but does not address the style choice of having `get_collection_path` be a class method.

I'm going to mark this as resolved, but let me know if the class method must be implemented!

Reply:
This works for me!